File: julia/src/bm.jl
using Random
using ThreadsX

include("utils.jl")

struct Bm{T<:AbstractFloat}
    x₀::T
    D::T
    function Bm{T}(x₀::T=0.0, D::T=0.5) where T<:AbstractFloat
        new{T}(x₀, D)
    end
end

function simulate(bm::Bm{T}, duration::T, τ::T=0.01) where T<:AbstractFloat
    σ = sqrt(2 * bm.D * τ)
    t = collect(zero(T):τ:duration)
    x = Vector{T}(undef, length(t))
    n = length(t) - 1
    ξ = randn(T, n - 1)
    xₙ = bm.x₀
    @inbounds for i in 1:(n-1)
        xₙ += ξ[i] * σ
        x[i] = xₙ
    end
    last_step = duration - (n - 1) * τ
    noise = randn(T) * sqrt(2 * bm.D * last_step)
    xₙ += noise
    x[end] = xₙ
    t, x
end

function displacement(bm::Bm{T}, duration::T, τ::T=0.01) where T<:AbstractFloat
    n = ceil(Int, duration / τ)
    σ = sqrt(2 * bm.D * τ)
    ξ = randn(T, n - 1)
    Δx = zero(T)
    @inbounds for i in 1:(n-1)
        Δx += ξ[i] * σ
    end
    last_step = duration - (n - 1) * τ
    Δx += randn(T) * sqrt(2 * bm.D * last_step)
    Δx
end


function msd(bm::Bm{T}, duration::T, N::Int=10_000, τ::T=0.01) where T<:AbstractFloat
    displacements = ThreadsX.map(1:N) do _
        displacement(bm, duration, τ)
    end
    sum(displacements .^ 2) / N
end

bm = Bm{Float64}()

simulate(bm, 1.0, 0.01)
msd(bm, 1.0, 100, 0.01)

bench("Brownian motion simulation", () -> simulate(bm, 100.0, 0.01))
bench("Brownian motion msd", () -> msd(bm, 100.0, 10_000, 0.01))

File: python/bm.py
from math import ceil, sqrt

import numpy as np
from numba import (  # pyright: ignore [reportMissingTypeStubs]
    njit,  # pyright: ignore [reportUnknownVariableType]
    prange,
)


class PyBm:
    def __init__(self, starting_point: float = 0.0, diffusivity: float = 0.5):
        self.starting_point: float = starting_point
        self.diffusivity: float = diffusivity

    def simulate(self, duration: float, time_step: float = 0.01):
        num_steps = ceil(duration / time_step)
        sigma = sqrt(2 * self.diffusivity * time_step)
        noises = np.random.standard_normal(num_steps) * sigma
        last_step = duration - (num_steps - 1) * time_step
        noises[-1] = noises[-1] * sqrt(last_step / time_step)
        noises = np.insert(noises, 0, self.starting_point)
        t = np.arange(0, duration, time_step)
        if t[-1] != duration:
            t = np.append(t, duration)
        x = np.cumsum(noises)
        return t, x

    def displacement(self, duration: float, time_step: float = 0.01):
        num_steps = ceil(duration / time_step)
        sigma = sqrt(2 * self.diffusivity * time_step)
        noises = np.random.standard_normal(num_steps) * sigma
        last_step = duration - (num_steps - 1) * time_step
        noises[-1] = noises[-1] * sqrt(last_step / time_step)
        return np.sum(noises)

    def msd(
        self,
        duration: float,
        N: int = 10_000,
        tau: float = 0.01,
    ):
        return _msd(
            duration,
            self.diffusivity,
            N,
            tau,
        )


@njit(parallel=True)  # pyright: ignore [reportUntypedFunctionDecorator]
def _msd(
    duration: float,
    diffusivity: float = 0.5,
    N: int = 10_000,
    time_step: float = 0.01,
):
    total = 0.0
    for _ in prange(N):
        num_steps = ceil(duration / time_step)
        sigma = sqrt(2 * diffusivity * time_step)
        noises = np.random.standard_normal(num_steps) * sigma
        last_step = duration - (num_steps - 1) * time_step
        noises[-1] = noises[-1] * sqrt(last_step / time_step)
        total += np.sum(noises) ** 2
    return total / N


if __name__ == "__main__":
    from utils import bench

    bench_size = 100

    duration = 100.0
    time_step = 0.01
    particles = 10000

    py_bm = PyBm()

    def py_simulate():
        return py_bm.simulate(duration, time_step)

    def py_msd():
        return py_bm.msd(duration, particles, time_step)

    bench("Brownian motion simulation", py_simulate, bench_size)
    bench("Brownian motion msd", py_msd, bench_size)

    from diffusionx.simulation import Bm

    rs_bm = Bm()

    def rs_simulate():
        return rs_bm.simulate(duration, time_step)

    def rs_msd():
        return rs_bm.msd(duration, particles=particles, time_step=time_step)

    bench("Brownian motion simulation (PyO3)", rs_simulate, bench_size)
    bench("Brownian motion msd (PyO3)", rs_msd, bench_size)
